DigitalLearningSolutions.Web.Tests/Controllers/TrackingSystem/CourseSetup/ManageCourseControllerTests.cs

```diff
@@ -49,5 +51,78 @@ public void Index_returns_ManageCourse_page_when_appropriate_course_found()
             // Then
             result.Should().BeViewResult().WithDefaultViewName().ModelAs<ManageCourseViewModel>();
         }
+
+        [Test]
+        public void SaveLearningPathwayDefaults_save_calls_correct_method()
+        {
+            // Given
+            var model = new EditLearningPathwayDefaultsViewModel(1, "6", "12", false, false);
+
+            A.CallTo(
+                () => courseService.UpdateLearningPathwayDefaultsForCourse(
+                    A<int>._,
+                    A<int>._,
+                    A<int>._,
+                    A<bool>._,
+                    A<bool>._
+                )
+            ).DoesNothing();
+
+            // When
+            var result = controller.SaveLearningPathwayDefaults(1, model);
+
+            // Then
+            A.CallTo(
+                () => courseService.UpdateLearningPathwayDefaultsForCourse(
+                    1,
+                    6,
+                    12,
+                    false,
+                    false
+                )
+            ).MustHaveHappened();
+            result.Should().BeRedirectToActionResult().WithActionName("Index");
+        }
+
+        [Test]
+        public void SaveLearningPathwayDefaults_saves_if_number_input_is_null()
+        {
+            // Given
+            var model = new EditLearningPathwayDefaultsViewModel(1, null, null, false, false);
+
+            // When
+            var result = controller.SaveLearningPathwayDefaults(1, model);
+
+            // Then
+            result.Should().BeRedirectToActionResult().WithActionName("Index");
+        }
+
+        [Test]
+        public void SaveLearningPathwayDefaults_shows_validation_errors_if_number_input_contains_non_numbers()
+        {
+            // Given
+            var model = new EditLearningPathwayDefaultsViewModel(1, "9.0", "asdfg", false, false);
+
+            // When
+            var result = controller.SaveLearningPathwayDefaults(1, model);
+
+            // Then
+            result.Should().BeViewResult().ModelAs<EditLearningPathwayDefaultsViewModel>();
+            Assert.IsFalse(controller.ModelState.IsValid);
+        }
+
+        [Test]
+        public void SaveLearningPathwayDefaults_shows_validation_errors_if_number_input_is_outside_range()
+        {
+            // Given
+            var model = new EditLearningPathwayDefaultsViewModel(1, "-1", "49", false, false);
+
+            // When
+            var result = controller.SaveLearningPathwayDefaults(1, model);
+
+            // Then
+            result.Should().BeViewResult().ModelAs<EditLearningPathwayDefaultsViewModel>();
+            Assert.IsFalse(controller.ModelState.IsValid);

```

Comment:
Added that check to the validation tests in the controller, but after making the validation attribute all those validation tests failed. So, I moved them to a test file for the validation attribute itself, but just shout if you think there should still be validation tests in the controller and I can figure out what was going wrong.

Reply:
If I am understanding the situation correctly, I’d imagine the failing tests would be to do with the ModelState not being bound correctly with the controller. But that's only a guess. I can’t find any example of where we set this up correctly, so I’d say you can look into adding the model to the controller’s ModelState dictionary (probably similar to how we set up TempData in the ControllerContextHelper?), or you can just ignore the valid/invalid tests as we don’t seem to have anything like that elsewhere from what I can tell.

Reply:
I looked for an example of a controller that test validation with a NoWhitespaceAttribute applied to a view model, and found that MyAccountController tests only for if the save method does not call the service if the model state is invalid, rather than testing any specific invalid inputs. So I added a similar test on the ManageCourseController :)